File: tests/test_v1_26_d_info_board_load_gate.py
#!/usr/bin/env python3
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
loader = (ROOT / 'public/js/memo-counter.js').read_text(encoding='utf-8')
version = (ROOT / 'app/version.php').read_text(encoding='utf-8')
runner = (ROOT / 'tests/run-current-features.sh').read_text(encoding='utf-8')

checks = []

def check(condition: bool, message: str) -> None:
    checks.append(bool(condition))
    print(('PASS' if condition else 'FAIL') + ': ' + message)

check("widget.infoboard.fetch" in loader and "function isInfoBoardFetch" in loader,
      'load gate targets only Information Board aggregate fetch requests')
check("return originalAjax.apply(this, arguments)" in loader,
      'normal RSS, Weather and other Ajax requests continue immediately')
check("Number($.active || 0)" in loader and "IDLE_CHECKS_REQUIRED = 2" in loader,
      'Information Board waits for a stably idle Dashboard Ajax queue')
check("idleChecks = 0" in loader and "idleChecks++" in loader,
      'new Dashboard network activity resets the idle window')
check("var queue = []" in loader and "var running = false" in loader,
      'Information Board fetches are queued and limited to one running request')
check("xhr.done(function ()" in loader and ".fail(function ()" in loader and ".always(function ()" in loader,
      'queued request preserves the existing jqXHR done/fail/always contract')
check("proxy.abort = function" in loader,
      'queued Information Board request remains abortable')
check("APP_ASSET_REVISION = '1.25.0'" in version,
      'global asset revision remains at the formal V1.25 baseline during V1.26 development')
check("APP_VERSION = '1.25.0'" in version,
      'formal app version remains unchanged during V1.26 development')
check('test_v1_26_d_info_board_load_gate.py' in runner,
      'active current-feature runner executes the gate static contract')
check('test_v1_26_d_info_board_load_gate.js' in runner,
      'active current-feature runner executes the gate runtime helper test')

failed = len(checks) - sum(checks)
print(f'RESULT: PASS {len(checks) - failed} / FAIL {failed} / SKIP 0')
if failed:
    raise SystemExit(1)

File: tests/test_v1_26_d_info_board_load_gate.js
'use strict';

const fs = require('fs');
const path = require('path');
const vm = require('vm');

const root = path.resolve(__dirname, '..');
const source = fs.readFileSync(path.join(root, 'public/js/memo-counter.js'), 'utf8');

let ajaxCalls = 0;
function originalAjax(options) {
    ajaxCalls += 1;
    return {options: options};
}

function Deferred() {
    return {
        promise: function () { return {}; },
        resolveWith: function () {},
        rejectWith: function () {}
    };
}

const jQueryStub = {
    ajax: originalAjax,
    Deferred: Deferred,
    active: 0
};
const windowStub = {
    jQuery: jQueryStub,
    setTimeout: function () { return 1; },
    clearTimeout: function () {}
};
const sandbox = {
    window: windowStub,
    Number: Number,
    String: String,
    Array: Array,
    Object: Object,
    Math: Math,
    setTimeout: windowStub.setTimeout,
    clearTimeout: windowStub.clearTimeout
};

vm.runInNewContext(source, sandbox, {filename: 'memo-counter.js'});

const gate = windowStub.RssInfoBoardAjaxGate;
if (!gate) {
    throw new Error('RssInfoBoardAjaxGate was not exported');
}

const checks = [];
function check(condition, message) {
    checks.push(Boolean(condition));
    console.log((condition ? 'PASS' : 'FAIL') + ': ' + message);
}

check(gate.isInfoBoardFetch({data: {action: 'widget.infoboard.fetch'}}) === true,
      'Information Board fetch is recognized');
check(gate.isInfoBoardFetch({data: {action: 'feed.fetch'}}) === false,
      'normal RSS feed fetch is not gated');
check(gate.isInfoBoardFetch({data: {action: 'widget.weather.fetch'}}) === false,
      'other remote widget fetch is not gated');
check(gate.canStart(4, 10) === false, 'active Dashboard requests block Information Board');
check(gate.canStart(1, 2) === false, 'one active Dashboard request still blocks Information Board');
check(gate.canStart(0, 1) === false, 'one idle poll is not enough');
check(gate.canStart(0, 2) === true, 'two stable idle polls allow Information Board');

const normalResult = jQueryStub.ajax({data: {action: 'feed.fetch'}});
check(ajaxCalls === 1 && normalResult && normalResult.options.data.action === 'feed.fetch',
      'normal RSS request reaches the original Ajax implementation immediately');

jQueryStub.ajax({data: {action: 'widget.infoboard.fetch'}});
check(ajaxCalls === 1 && gate.pendingCount() === 1,
      'Information Board request is queued instead of competing immediately');

const failed = checks.length - checks.filter(Boolean).length;
console.log('RESULT: PASS ' + (checks.length - failed) + ' / FAIL ' + failed + ' / SKIP 0');
if (failed > 0) {
    process.exit(1);
}
